batcher/src/lib.rs

```diff
@@ -31,275 +28,229 @@ pub mod types;
-        future::select(get_incoming, receive_from_others).await;
+        pin_mut!(get_incoming, send_outgoing);
+        future::select(get_incoming, send_outgoing).await;
 
         info!("{} disconnected", &addr);
     }
 
     async fn handle_message(
-        self: &Arc<Self>,
-        tx: UnboundedSender<Message>,
+        self: Arc<Self>,
         message: Message,
     ) -> Result<(), tokio_tungstenite::tungstenite::Error> {
-        // TODO: Handle errors
-
         // Deserialize task from message
         let verification_data: VerificationData =
             serde_json::from_str(message.to_text().expect("Message is not text"))
                 .expect("Failed to deserialize task");
 
-        let proof = verification_data.proof.as_slice();
-        let vm_program_code = verification_data.vm_program_code.as_ref();
-
-        let response = match verification_data.proving_system {
-            types::ProvingSystemId::SP1 => {
-                let elf = vm_program_code.expect("VM program code is required");
-
-                let elf = elf.as_slice();
-
-                self.verify_sp1_proof(proof, elf)
-            }
-            types::ProvingSystemId::GnarkPlonkBls12_381
-            | types::ProvingSystemId::GnarkPlonkBn254
-            | types::ProvingSystemId::Groth16Bn254 => {
-                let vk = verification_data
-                    .verification_key
-                    .as_ref()
-                    .expect("Verification key is required");
-
-                let public_inputs = verification_data
-                    .pub_input
-                    .as_ref()
-                    .expect("Public input is required");
-
-                let is_valid =
-                    verify_gnark(&verification_data.proving_system, proof, public_inputs, vk);
-
-                debug!("Proof is valid: {}", is_valid);
-
-                if is_valid {
-                    Ok(())
-                } else {
-                    Err(anyhow::anyhow!("Failed to verify proof"))
-                }
-            }
-        };
-
-        let response = match response {
-            Ok(_) => {
-                let task_bytes = bincode::serialize(&verification_data)
-                    .expect("Failed to bincode serialize task");
-
-                self.add_task(verification_data).await;
-
-                let task_bytes = Bytes::from(task_bytes);
-                let mut hasher = Sha3_256::new();
-                hasher.update(&task_bytes);
-                let hash = hasher.finalize().to_vec();
-
-                Ok(hash)
-            }
-            Err(e) => Err(e.to_string()),
+        if verification_data.verify() {
+            self.add_to_batch(verification_data).await;
+        // FIXME(marian): This condition should be removed once we have FFI bindings with Gnark.
+        } else if verification_data.proving_system == ProvingSystemId::Groth16Bn254 {
+            self.add_to_batch(verification_data).await;
```

Comment:
This condition can already be removed right

```suggestion
```

Reply:
good catch, fixed 13e9c56dc993d19b17b088fd1d3b4b2d1920b1a1